util/vizerror/vizerror.go

```diff
@@ -22,25 +23,51 @@ func (e Error) Error() string {
 
 // New returns an error that formats as the given text. It always returns a vizerror.Error.
 func New(text string) error {
-	return Error{errors.New(text)}
+	err := errors.New(text)
+	return Error{err: err, wrapped: err}
 }
 
 // Errorf returns an Error with the specified format and values. It always returns a vizerror.Error.
 func Errorf(format string, a ...any) error {
-	return Error{fmt.Errorf(format, a...)}
+	err := fmt.Errorf(format, a...)
+	return Error{err: err, wrapped: err}
 }
 
 // Unwrap returns the underlying error.
 func (e Error) Unwrap() error {
-	return e.err
+	return e.wrapped
 }
 
 // Wrap wraps err with a vizerror.Error.
+//
+// Deprecated: this is almost always the wrong thing to do. Are you really sure
+// you know exactly what err.Error() will stringify to and be safe to show to
+// users? [WrapWithMessage] is probably what you want.
 func Wrap(err error) error {
 	if err == nil {
 		return nil
 	}
-	return Error{err}
+	return Error{err: err, wrapped: err}
+}
+
+// WrapWithMessage wraps the given error with a msg that's safe to display
+// to end users. The text of the wrapped error will not be displayed to end
+// users.
+//
+// WrapWithMessage should almost always be preferred to [Wrap].
+func WrapWithMessage(err error, msg string) error {
+	return Error{
+		err:     errors.New(msg),
+		wrapped: err,
+	}
+}
+
+// WrapWithMessageF is like [WrapWithMessage] but allows string formatting.
+func WrapWithMessageF(err error, format string, a ...any) error {
+	return Error{
+		err:     fmt.Errorf(format, a...),
+		wrapped: err,
+	}
```

Comment:
let's drop this method. Callers can do their own Sprintf if they need a complicated public message.

Plus this adds ambiguity into what what tree of errors Unwrap looks at if the caller uses %w here.